### PR TITLE
Move channel logic into `Chan`

### PR DESCRIPTION
**Pull request by thomaseizinger**
As part of trying to come up with an alternative to https://github.com/Restioson/xtra/pull/120, I started to move a bit of code around and ended up with this patch set!

Each commit in its own builds and has passing tests and I was careful to try and not alter any behaviour. In other words, this is a pure refactoring.

The goal was as follows:

- Have `Chan` house the core channel implementation with descriptive function names for key operations
- Have `Sender` and `Receiver` be thin layers on top that combine `Chan` with reference counting and provide a `Future`-based APIs for sending and receiving

I can extract some of these commits into individual PRs if you want. The diff will be quite hard to follow once it is squashed.

### CODE REVIEW

**Review by Restioson — APPROVED** [review]

**Comment by thomaseizinger** [comment]
I just noticed absolute irony. This PR starts with a commit that removes `Sender::send` and ends with a commit that adds it again. I guess I've gone full circle 😂
